File: feed-wrangler/lib/feed_wrangler/cert_forge.ex
defmodule FeedWrangler.CertForge do
  @moduledoc """
  cert-forge client for feed-wrangler.

  Obtains instance cert from cert-forge on startup.
  Private key held in memory — never written to the shared volume.
  Uses curl for all HTTP calls — avoids adding HTTP client dependencies.
  """
  require Logger

  @service_name "feed-wrangler"
  @endpoint     "https://feed-wrangler:4007"
  @enrollment_cert System.get_env("ENROLLMENT_CERT", "/certs/enrollment.crt")
  @enrollment_key  System.get_env("ENROLLMENT_KEY",  "/certs/enrollment.key")

  defstruct [:ca_cert, :instance_cert, :instance_key, :fingerprint, :instance_cn, :instance_id]

  @type t :: %__MODULE__{
    ca_cert:      binary(),
    instance_cert: binary(),
    instance_key:  binary(),
    fingerprint:  String.t(),
    instance_cn:  String.t(),
    instance_id:  String.t(),
  }

  @doc "Obtain all cert material — call once on startup, returns {:ok, mat} or {:error, reason}"
  def obtain_certs do
    forge_url    = System.get_env("CERT_FORGE_URL", "https://cert-forge:4014")
    public_port  = System.get_env("PUBLIC_PORT", "4016")
    enroll_port  = System.get_env("ENROLLMENT_PORT", "4015")
    instance_id  = System.get_env("HOSTNAME", "local")

    public_url  = derive_url(forge_url, public_port) |> String.replace("https://", "http://")
    enroll_url  = derive_url(forge_url, enroll_port)

    with {:ok, ca_cert}    <- fetch_ca_cert(public_url),
         {:ok, cert, key, fp, cn} <- request_instance_cert(enroll_url, instance_id) do
      mat = %__MODULE__{
        ca_cert:       ca_cert,
        instance_cert: cert,
        instance_key:  key,
        fingerprint:   fp,
        instance_cn:   cn,
        instance_id:   instance_id,
      }
      Logger.info("[cert-forge] Instance cert obtained (CN=#{cn})")
      {:ok, mat}
    end
  end

  defp derive_url(url, port) do
    Regex.replace(~r/:\d+$/, url, ":#{port}")
  end

  defp fetch_ca_cert(public_url, attempt \\ 1)
  defp fetch_ca_cert(_, 31), do: {:error, "Could not obtain CA cert after 30 attempts"}
  defp fetch_ca_cert(public_url, attempt) do
    url = String.trim_trailing(public_url, "/") <> "/ca"
    {output, code} = System.cmd("curl", ["--silent", "--fail", url], stderr_to_stdout: true)
    if code == 0 do
      case JSON.decode(output) do
        {:ok, %{"ca_cert" => ca_cert}} ->
          Logger.info("[cert-forge] CA cert obtained")
          {:ok, ca_cert}
        _ ->
          Logger.info("[cert-forge] Waiting for /ca (attempt #{attempt}/30)")
          Process.sleep(2_000)
          fetch_ca_cert(public_url, attempt + 1)
      end
    else
      Logger.info("[cert-forge] Waiting for /ca (attempt #{attempt}/30): curl exit #{code}")
      Process.sleep(2_000)
      fetch_ca_cert(public_url, attempt + 1)
    end
  end

  defp request_instance_cert(enroll_url, instance_id, attempt \\ 1)
  defp request_instance_cert(_, _, 31), do: {:error, "Could not obtain instance cert after 30 attempts"}
  defp request_instance_cert(enroll_url, instance_id, attempt) do
    url  = String.trim_trailing(enroll_url, "/") <> "/instance-cert"
    body = JSON.encode!(%{service_name: @service_name, instance_id: instance_id})

    enroll_cert = System.get_env("ENROLLMENT_CERT", "/certs/enrollment.crt")
    enroll_key  = System.get_env("ENROLLMENT_KEY",  "/certs/enrollment.key")

    {output, code} = System.cmd("curl", [
      "--silent", "--fail",
      "--cert",    enroll_cert,
      "--key",     enroll_key,
      "--insecure",  # enrollment CA verified server-side
      "--request", "POST",
      "--header",  "Content-Type: application/json",
      "--data",    body,
      url
    ], stderr_to_stdout: true)

    if code == 0 do
      case JSON.decode(output) do
        {:ok, %{"cert" => cert, "key" => key, "fingerprint" => fp, "instance_cn" => cn}} ->
          {:ok, cert, key, fp, cn}
        _ ->
          Logger.info("[cert-forge] Waiting for /instance-cert (attempt #{attempt}/30)")
          Process.sleep(2_000)
          request_instance_cert(enroll_url, instance_id, attempt + 1)
      end
    else
      Logger.info("[cert-forge] Waiting for /instance-cert (attempt #{attempt}/30): curl exit #{code}")
      Process.sleep(2_000)
      request_instance_cert(enroll_url, instance_id, attempt + 1)
    end
  end

  @doc "Sign payload via cert-forge /sign, returns {:ok, signature} or {:error, reason}"
  def sign_payload(mat, payload) do
    forge_url  = System.get_env("CERT_FORGE_URL", "https://cert-forge:4014")
    encoded    = Base.encode64(payload)
    body       = JSON.encode!(%{
      service_name: @service_name,
      instance_id:  mat.instance_id,
      payload:      encoded,
    })

    # Write cert+key to temp files for curl
    with {:ok, cert_path, key_path, ca_path} <- write_temp_certs(mat),
         {output, 0} <- System.cmd("curl", [
           "--silent", "--fail",
           "--cacert",  ca_path,
           "--cert",    cert_path,
           "--key",     key_path,
           "--request", "POST",
           "--header",  "Content-Type: application/json",
           "--data",    body,
           forge_url <> "/sign"
         ], stderr_to_stdout: true),
         {:ok, %{"signature" => sig}} <- JSON.decode(output) do
      cleanup_temp_files([cert_path, key_path, ca_path])
      {:ok, sig}
    else
      {output, code} ->
        Logger.warning("[cert-forge] /sign failed (exit=#{code}): #{output}")
        {:error, "sign failed"}
      err ->
        {:error, inspect(err)}
    end
  end

  @doc "Register with Augur Canis"
  def self_register(mat) do
    ac_url    = System.get_env("AUGUR_CANIS_URL", "https://augur-canis:4010")
    timestamp = DateTime.utc_now() |> DateTime.to_iso8601()
    payload   = @service_name <> @endpoint <> mat.fingerprint <> timestamp

    do_register(mat, ac_url, timestamp, payload, 1)
  end

  defp do_register(_, _, _, _, 11) do
    Logger.warning("[feed-wrangler] selfRegister: giving up after 10 attempts")
  end
  defp do_register(mat, ac_url, timestamp, payload, attempt) do
    case sign_payload(mat, payload) do
      {:ok, sig} ->
        body = JSON.encode!(%{
          service_name:     @service_name,
          network_endpoint: @endpoint,
          cert_fingerprint: mat.fingerprint,
          cert_pem:         mat.instance_cert,
          timestamp:        timestamp,
          signature:        sig,
        })
        url = String.trim_trailing(ac_url, "/") <> "/services/register"
        with {:ok, cert_path, key_path, ca_path} <- write_temp_certs(mat),
             {response, 0} <- System.cmd("curl", [
               "--silent",
               "--cacert",  ca_path,
               "--cert",    cert_path,
               "--key",     key_path,
               "--request", "POST",
               "--header",  "Content-Type: application/json",
               "--data",    body,
               url
             ], stderr_to_stdout: true),
             {:ok, ack} <- JSON.decode(response) do
          cleanup_temp_files([cert_path, key_path, ca_path])
          Logger.info("[feed-wrangler] selfRegister: registered with AC (status=#{ack["status"]})")
        else
          _ ->
            Logger.info("[feed-wrangler] selfRegister: attempt #{attempt}/10 failed")
            Process.sleep(3_000)
            do_register(mat, ac_url, timestamp, payload, attempt + 1)
        end
      {:error, reason} ->
        Logger.info("[feed-wrangler] selfRegister: sign failed (attempt #{attempt}/10): #{reason}")
        Process.sleep(3_000)
        do_register(mat, ac_url, timestamp, payload, attempt + 1)
    end
  end

  # Write cert+key+ca to temp files for curl — cleaned up immediately after use
  defp write_temp_certs(mat) do
    cert_path = "/tmp/fw-cert-#{:erlang.unique_integer([:positive])}.crt"
    key_path  = "/tmp/fw-key-#{:erlang.unique_integer([:positive])}.key"
    ca_path   = "/tmp/fw-ca-#{:erlang.unique_integer([:positive])}.crt"
    File.write!(cert_path, mat.instance_cert)
    File.write!(key_path,  mat.instance_key)
    File.write!(ca_path,   mat.ca_cert)
    {:ok, cert_path, key_path, ca_path}
  end

  defp cleanup_temp_files(paths) do
    Enum.each(paths, &File.rm/1)
  end
end

File: gateway/certforge.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// cert-forge client
//
// Every service calls obtainCerts() on startup instead of reading cert
// files from /certs/. The private key is delivered over the encrypted
// enrollment mTLS connection and held in memory only — never on disk,
// never on the shared volume.
//
// Flow:
//  1. GET /ca (unauthenticated) — get CA cert
//  2. POST /instance-cert on enrollment port (enrollment mTLS) — get cert+key
//  3. Hold both in memory, build TLS configs from them
//  4. All signing delegated to cert-forge /sign — algorithm is cert-forge's concern
// ---------------------------------------------------------------------------

type CertMaterial struct {
	CACertPEM    []byte
	InstanceCert tls.Certificate
	Fingerprint  string
	InstanceCN   string
	InstanceID   string
	ServiceName  string
	ValidUntil   time.Time
}

func obtainCerts(serviceName string) *CertMaterial {
	forgeURL   := cfEnvOr("CERT_FORGE_URL", "https://cert-forge:4014")
	enrollPort := cfEnvOr("ENROLLMENT_PORT", "4015")
	enrollCert := cfEnvOr("ENROLLMENT_CERT", "/certs/enrollment.crt")
	enrollKey  := cfEnvOr("ENROLLMENT_KEY", "/certs/enrollment.key")
	instanceID := os.Getenv("HOSTNAME")
	if instanceID == "" {
		instanceID = "local"
	}

	enrollURL := deriveEnrollmentURL(forgeURL, enrollPort)

	caCertPEM := fetchCACert(forgeURL)
	sans      := buildSANs(serviceName, instanceID)
	certPEM, keyPEM, fingerprint, instanceCN, validUntil :=
		requestInstanceCert(enrollURL, enrollCert, enrollKey, serviceName, instanceID, sans)

	tlsCert, err := tls.X509KeyPair(certPEM, keyPEM)
	if err != nil {
		log.Fatalf("[%s] Failed to parse instance cert+key: %v", serviceName, err)
	}

	log.Printf("[%s] Instance cert obtained from cert-forge (CN=%s, valid until %s)",
		serviceName, instanceCN, validUntil.Format(time.RFC3339))

	return &CertMaterial{
		CACertPEM:    caCertPEM,
		InstanceCert: tlsCert,
		Fingerprint:  fingerprint,
		InstanceCN:   instanceCN,
		InstanceID:   instanceID,
		ServiceName:  serviceName,
		ValidUntil:   validUntil,
	}
}

// buildSANs constructs the SAN list for the gateway's instance cert.
// Because cert-forge's sans field overrides (not merges) the defaults,
// we must explicitly include the standard entries alongside any
// environment-supplied external hostname.
func buildSANs(serviceName, instanceID string) []string {
	sans := []string{
		serviceName,
		fmt.Sprintf("%s-%s", serviceName, instanceID),
		"localhost",
	}
	if h := os.Getenv("GATEWAY_EXTERNAL_HOSTNAME"); h != "" {
		sans = append(sans, h)
	}
	return sans
}

func deriveEnrollmentURL(forgeURL, enrollPort string) string {
	parts := strings.Split(forgeURL, ":")
	if len(parts) == 3 {
		return parts[0] + ":" + parts[1] + ":" + enrollPort
	}
	return strings.TrimRight(forgeURL, "/") + ":" + enrollPort
}

func fetchCACert(forgeURL string) []byte {
	// /ca is on plain HTTP — no TLS, no client cert needed
	publicPort := cfEnvOr("PUBLIC_PORT", "4016")
	publicURL := deriveEnrollmentURL(forgeURL, publicPort)
	publicURL = strings.Replace(publicURL, "https://", "http://", 1)
	url := strings.TrimRight(publicURL, "/") + "/ca"
	client := &http.Client{Timeout: 5 * time.Second}

	for attempt := 1; attempt <= 30; attempt++ {
		resp, err := client.Get(url)
		if err != nil {
			log.Printf("[certforge] Waiting for cert-forge /ca (attempt %d/30): %v", attempt, err)
			time.Sleep(2 * time.Second)
			continue
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()

		var result map[string]string
		if err := json.Unmarshal(body, &result); err != nil || result["ca_cert"] == "" {
			log.Printf("[certforge] Invalid /ca response (attempt %d/30)", attempt)
			time.Sleep(2 * time.Second)
			continue
		}
		log.Printf("[certforge] CA cert obtained")
		return []byte(result["ca_cert"])
	}
	log.Fatalf("[certforge] Could not obtain CA cert after 30 attempts")
	return nil
}

func requestInstanceCert(enrollURL, enrollCertPath, enrollKeyPath, serviceName, instanceID string, sans []string) ([]byte, []byte, string, string, time.Time) {
	enrollTLSCert, err := tls.LoadX509KeyPair(enrollCertPath, enrollKeyPath)
	if err != nil {
		log.Fatalf("[certforge] Cannot load enrollment cert %s: %v", enrollCertPath, err)
	}

	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{
				Certificates:       []tls.Certificate{enrollTLSCert},
				InsecureSkipVerify: true,
				MinVersion:         tls.VersionTLS13,
			},
		},
		Timeout: 10 * time.Second,
	}

	reqBody, err := json.Marshal(map[string]interface{}{
		"service_name": serviceName,
		"instance_id":  instanceID,
		"sans":         sans,
	})
	if err != nil {
		log.Fatalf("[certforge] Failed to marshal instance-cert request: %v", err)
	}

	url := strings.TrimRight(enrollURL, "/") + "/instance-cert"

	for attempt := 1; attempt <= 30; attempt++ {
		req, _ := http.NewRequest(http.MethodPost, url, strings.NewReader(string(reqBody)))
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			log.Printf("[certforge] Waiting for /instance-cert (attempt %d/30): %v", attempt, err)
			time.Sleep(2 * time.Second)
			continue
		}

		if resp.StatusCode != http.StatusOK {
			io.Copy(io.Discard, resp.Body)
			resp.Body.Close()
			log.Printf("[certforge] /instance-cert returned %d (attempt %d/30)", resp.StatusCode, attempt)
			time.Sleep(2 * time.Second)
			continue
		}

		respBody, _ := io.ReadAll(resp.Body)
		resp.Body.Close()

		var result map[string]interface{}
		if err := json.Unmarshal(respBody, &result); err != nil {
			log.Printf("[certforge] Invalid /instance-cert response: %v", err)
			time.Sleep(2 * time.Second)
			continue
		}

		certPEM  := []byte(result["cert"].(string))
		keyPEM   := []byte(result["key"].(string))
		fp       := result["fingerprint"].(string)
		cn       := result["instance_cn"].(string)
		vuStr, _ := result["valid_until"].(string)
		vu, _    := time.Parse(time.RFC3339, vuStr)

		return certPEM, keyPEM, fp, cn, vu
	}
	log.Fatalf("[certforge] Could not obtain instance cert after 30 attempts")
	return nil, nil, "", "", time.Time{}
}

func buildServerTLS(mat *CertMaterial) *tls.Config {
	caPool := x509.NewCertPool()
	caPool.AppendCertsFromPEM(mat.CACertPEM)
	return &tls.Config{
		Certificates: []tls.Certificate{mat.InstanceCert},
		ClientCAs:    caPool,
		ClientAuth:   tls.RequireAndVerifyClientCert,
		MinVersion:   tls.VersionTLS13,
	}
}

func buildClientTLS(mat *CertMaterial) *tls.Config {
	caPool := x509.NewCertPool()
	caPool.AppendCertsFromPEM(mat.CACertPEM)
	return &tls.Config{
		Certificates: []tls.Certificate{mat.InstanceCert},
		RootCAs:      caPool,
		MinVersion:   tls.VersionTLS13,
	}
}

// signPayload delegates signing to cert-forge /sign.
// The calling service never handles the private key for signing —
// cert-forge owns the algorithm and the key material for this purpose.
func signPayload(mat *CertMaterial, payload []byte) (string, error) {
	forgeURL := cfEnvOr("CERT_FORGE_URL", "https://cert-forge:4014")

	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: buildClientTLS(mat),
		},
		Timeout: 10 * time.Second,
	}

	body := fmt.Sprintf(`{"service_name":%q,"instance_id":%q,"payload":%q}`,
		mat.ServiceName, mat.InstanceID, base64.StdEncoding.EncodeToString(payload))

	req, _ := http.NewRequest(http.MethodPost,
		strings.TrimRight(forgeURL, "/")+"/sign",
		strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("sign request failed: %v", err)
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("cert-forge /sign returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result map[string]string
	if err := json.Unmarshal(respBody, &result); err != nil {
		return "", fmt.Errorf("sign response invalid: %v", err)
	}
	sig := result["signature"]
	if sig == "" {
		return "", fmt.Errorf("cert-forge returned empty signature")
	}
	return sig, nil
}

// selfRegisterWithAC handles the full self-registration flow:
// sign the payload via cert-forge, POST to Augur Canis.
// Retries with backoff — AC may not be ready when this is first called.
func selfRegisterWithAC(mat *CertMaterial, networkEndpoint string) {
	acURL := cfEnvOr("AUGUR_CANIS_URL", "https://augur-canis:4010")

	for attempt := 1; ; attempt++ {
		timestamp := time.Now().UTC().Format(time.RFC3339)
		payload   := mat.ServiceName + networkEndpoint + mat.Fingerprint + timestamp

		sig, err := signPayload(mat, []byte(payload))
		if err != nil {
			log.Printf("[%s] selfRegister: signing failed (attempt %d): %v", mat.ServiceName, attempt, err)
			time.Sleep(3 * time.Second)
			continue
		}

		var certPEM string
		if len(mat.InstanceCert.Certificate) > 0 {
			certPEM = string(pem.EncodeToMemory(&pem.Block{
				Type:  "CERTIFICATE",
				Bytes: mat.InstanceCert.Certificate[0],
			}))
		}

		body, _ := json.Marshal(map[string]interface{}{
			"service_name":     mat.ServiceName,
			"network_endpoint": networkEndpoint,
			"cert_fingerprint": mat.Fingerprint,
			"cert_pem":         certPEM,
			"timestamp":        timestamp,
			"signature":        sig,
		})

		client := &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: buildClientTLS(mat),
			},
			Timeout: 10 * time.Second,
		}

		req, _ := http.NewRequest(http.MethodPost,
			strings.TrimRight(acURL, "/")+"/services/register",
			strings.NewReader(string(body)))
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			log.Printf("[%s] selfRegister: AC unreachable (attempt %d): %v", mat.ServiceName, attempt, err)
			time.Sleep(3 * time.Second)
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode == http.StatusOK {
			var ack map[string]interface{}
			json.NewDecoder(resp.Body).Decode(&ack)
			log.Printf("[%s] selfRegister: registered with AC (status=%s)", mat.ServiceName, ack["status"])
			return
		}
		log.Printf("[%s] selfRegister: AC returned %d (attempt %d)", mat.ServiceName, resp.StatusCode, attempt)
		time.Sleep(3 * time.Second)
	}
}

func cfEnvOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
